### PR TITLE
Moving a container in a cluster keeping the same name

### LINKED ISSUE #4546: Fix same-name container move in cluster
# Required information

 * Distribution: Ubuntu
 * Distribution version: 8.04
 * The output of "lxc info" or if that fails:
```
root@lxd-dev-02:~# lxc info
config:
  core.https_address: 10.1.2.42:8443
  core.trust_password: true
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_stop_priority
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
- resources
- kernel_limits
- storage_api_volume_rename
- macaroon_authentication
- network_sriov
- console
- restrict_devlxd
- migration_pre_copy
- infiniband
- maas_network
- devlxd_events
- proxy
- network_dhcp_gateway
- file_get_symlink
- network_leases
- unix_device_hotplug
- storage_api_local_volume_handling
- operation_description
- clustering
- event_lifecycle
- storage_api_remote_volume_handling
- nvidia_runtime
api_status: stable
api_version: "1.0"
auth: trusted
public: false
auth_methods:
- tls
environment:
  addresses:
  - 10.1.2.42:8443
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
    MIIFTTCCAzWgAwIBAgIRAIYi9SOBaeGArc8ha+7ZyvIwDQYJKoZIhvcNAQELBQAw
    ODEcMBoGA1UEChMTbGludXhjb250YWluZXJzLm9yZzEYMBYGA1UEAwwPcm9vdEBs
    eGQtZGV2LTAxMB4XDTE4MDUwODIwNTU1N1oXDTI4MDUwNTIwNTU1N1owODEcMBoG
    A1UEChMTbGludXhjb250YWluZXJzLm9yZzEYMBYGA1UEAwwPcm9vdEBseGQtZGV2
    LTAxMIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEA0JxeMil0dOsHwO/x
    xIHJHXfC7VlNRP0WXNLb9glwUr3qVEExW+NG5XOdzts5uQm0wV1nhDrZoLr3AUwI
    dmrvREWbn7PQEr0Fw7ZHYPG1r+06b5m0fweEWuV6punvzZSCeiIV5M4w6S+93Xna
    C4fryXuIRIpMrXhhVzSclIGWiB4DA95KLN8oR8FfhEstZ6Jz+sYMMrOxDjbhEYLw
    OSV4MYgFhXlA+KGgeyCOrseiz9hycBeM8zniIxDzFIFTSGWd9B3omb4eK2li6zhb
    FNnNqsCdIOY7HsMJyD2bLJCKsuxXYkpN/XVJCNVOEoU13S1arTnIKPVzXvnFBpZx
    BREIbBQRI5fROYqlrbcMvOq/jEoWD9JQobK0OhfvdrTaJpWENF2ttf4QwANPx/zC
    DtSfIUbJ5I7iP0P+IuaShgwM2LG0aqyEwEpP3lPqV/LAPFiLj77QuxudiFVU0yZO
    iaA2UUFOskTNQNiMdZqD87j2Pp74dDrKSvmxTSwlXwJntUxAKoSUzaMBGhiRsuGp
    uc3atWRG+Ne7BMpepuyvEb2+kmFjexiyqH1pAXaAhNPJHithCiTFaOJQjbHx+n1n
    AxWqGnJQoDTyRm/JA1kyBo66NoDGW5X/RjVYIkD3Dc1svkYP6eWOB0Mi1mYj0qaq
    LG8jZ/e71ak9GSgdHDzXXvrLP0kCAwEAAaNSMFAwDgYDVR0PAQH/BAQDAgWgMBMG
    A1UdJQQMMAoGCCsGAQUFBwMBMAwGA1UdEwEB/wQCMAAwGwYDVR0RBBQwEoIKbHhk
    LWRldi0wMYcECgECKTANBgkqhkiG9w0BAQsFAAOCAgEAPTKfYKvM4tNuoItiIZ7C
    lKVN+73FriGg+xkl4APuT/FqciP4rXQ4eI+XW81BBc4jx1kK5ngmZbOO8tXROBZk
    lHp/6Vz1gxYLeLSbZA5Pi6dnJE5zJnPGrB3vUQgVxdlIF1zI2jtQ9h1h0wUH97nB
    uoK2f0Wd4S+z10bac6P4wGqrzwtFsfnT+077EIdpgtXeqQOxId1GaIfWEKzUhCRO
    Aj6aYw0bU5bigD0OhROI246nKk363X+4kZyMT7FNpUZMpESrw/ZjCBaCOkZkqOEq
    huCVn4yWhZ1sYIMlg9aitNeKaU70gOZDHajI+Lo8m2eemmiQoWGUqN8vu8Tk0Kzo
    s4bFhjPv/NFd9h2LvuulhBqmclPszgyhk9CrSPJL0K0GgcVUMG3IiuxoHBqVQn9M
    nPOVPlf2cYqb4OFC+kdDT3JqG29K+4am8gvV6rAFl+dCsRd6k45keWAB1MjQfl7N
    8whAL6TEXmiT1sb3pWvgd7E53SeIAXEPP/furs0excMJvu/yfZOGhaD/NTZzoC8g
    5WvNNxuuvuWISMJL0uHrn6NcgZ1rZFeOdMwOYewElyEkVLbFMpTacrGj3vYlIEfK
    MXSIiGOGMundoCvlDRRcLCqkYGbiu2Nlab23wpMWBxpFASsaZGgEXRRIn4ujie3t
    OboWBqb/G5F+82vh6wgV2t0=
    -----END CERTIFICATE-----
  certificate_fingerprint: fcc5ff4aea9e3513c7e2791e865bdae6f81b321c6c007a2fd7498d1d727060d0
  driver: lxc
  driver_version: 3.0.0
  kernel: Linux
  kernel_architecture: x86_64
  kernel_version: 4.15.0-20-generic
  server: lxd
  server_pid: 2728
  server_version: 3.0.0
  storage: zfs
  storage_version: 0.7.5-1ubuntu13
  server_clustered: true
  server_name: lxd-dev-02
```
# Issue description

I created 10 containers on a 3 node LXD cluster with ZFS storage back-end. I want to test migrating one of them from one node in the cluster but I get an error:

```
root@lxd-dev-02:~# lxc move cen5 --target lxd-dev-01
Error: Failed container creation: Container 'cen5' already exists
```

# Steps to reproduce

 1. Create a cluster with 3 nodes (for a proper quorum)
 2. Create a container anywhere in the cluster
 3. Run lxc move, using one of the node names in --target
```
root@lxd-dev-02:~# lxc move cen5 --target lxd-dev-03
Error: Failed container creation: Container 'cen5' already exists
```
other variants I have tried:
```
root@lxd-dev-02:~# lxc move cen5 cen5 --target lxd-dev-01
Error: Failed container creation: Container 'cen5' already exists
root@lxd-dev-02:~# lxc move cen5 lxd-dev-01:cen5
Error: The remote "lxd-dev-01" doesn't exist
root@lxd-dev-02:~# lxc move cen5 local:cen5
Error: already exists
root@lxd-dev-02:~# lxc move cen5 local:cen5 --target lxd-dev-01
Error: Failed container creation: Container 'cen5' already exists
root@lxd-dev-02:~# lxc move local:cen5 --target lxd-dev-01
Error: Failed container creation: Container 'cen5' already exists
```

I did discover that if I actually rename the container to something new it does attempt a migration (but gives me an entirely new error message.). If this is not allowed (migrating without renamaing) in a cluster configuration, then my suggestion is to no accept a single argument while using "--target"

### PR DESCRIPTION
**Pull request by freeekanayaka**
Fixes #4546.

Signed-off-by: Free Ekanayaka <free.ekanayaka@canonical.com>

### CODE REVIEW

**Comment by freeekanayaka** [comment]
For @stgraber: this is incomplete as it lacks handling of the "volatile" keys. Also, this is kind of an MVP since it uses existing APIs instead of creating a new one. I'd like to chat with you about whether this is acceptable (provided that we find a way to fix the volatile keys issue). If we do want the new API, I guess I could basically just copy to the API server the same logic in this branch? i.e. mostly what's currently in the cmdCopy.copyContainer helper.

**Comment by stgraber** [comment]
Sorry for only getting to this now. A few problems:
 - We really want this to be done server-side so that we don't need to replicate that kind of logic in every API client out there.
 - The temporary name approach will fail if the container name is at or close to the name length limit (63 chars)
 - As you mentioned, missing handling of volatile.apply_template

https://github.com/lxc/lxd/issues/4546#issuecomment-395122772

**Comment by freeekanayaka** [comment]
But should we factor out cmdCopy.copyContainer and use it both in the client? I don't see many other options.

**Comment by stgraber** [comment]
So what I expected here is:
 - lxd/container_post: Add logic so that if targetNode is set and container isn't CEPH backed we do:
    - GetContainer and save value of volatile.apply_template
    - Connect to source
    - Run CopyContainer using a temporary name (something like `move_<UUID>`) which wouldn't normally be allowed but can be allowed for this case by checking for the `lxd-cluster-notifier` User-Agent
    - Once that's done, DeleteContainer the original container
    - RenameContainer to the old name
    - Restore saved value of volatile.apply_template (unset if empty)
 - lxc/move: Change the logic to not special case CEPH in the cluster --target case but send all move requests within the cluster through that API

**Comment by stgraber** [comment]
Very little of the logic in copyContainer needs to make it to the server side as we won't be doing any changes to the container's config, devices, profiles, ... during the copy, so our call to CopyContainer should take the exact output of GetContainer and the only thing we need to set in args is the temporary name.

**Comment by freeekanayaka** [comment]
Should be ready for a new round. I implemented the change as you described, except that I'm naming the temporary container as "move-<UUID>", since such name is both a legal container name and a random enough one.

**Comment by stgraber** [comment]
jenkins: test this please

File: doc/rest-api.md [hunk 1]
```diff
@@ -725,7 +725,7 @@ Input:
         "ephemeral": true
     }
 
-### POST
+### POST (optional `?target=<node>`)
```
Comment (stgraber) [inline doc/rest-api.md]: s/node/member/

File: doc/rest-api.md [hunk 2]
```diff
@@ -739,7 +739,7 @@ Input (simple rename):
         "name": "new-name"
     }
 
-Input (migration across lxd instances):
+Input (migration across lxd instances or lxd cluster members):
```
Comment (stgraber) [inline doc/rest-api.md]: s/node/member/

File: doc/rest-api.md [hunk 3]
```diff
@@ -750,6 +750,8 @@ Input (migration across lxd instances):
 The migration does not actually start until someone (i.e. another lxd instance)
 connects to all the websockets and begins negotiation with the source.
 
+To migrate between cluster members the `?target=<node>` option is required.
```
Comment (stgraber) [inline doc/rest-api.md]: s/node/member/

File: lxc/move.go [hunk 4]
```diff
@@ -140,18 +140,18 @@ func (c *cmdMove) Run(cmd *cobra.Command, args []string) error {
...
 		}
-
-		if moved {
-			return nil
+		if c.flagContainerOnly {
+			return fmt.Errorf(i18n.G("The --container-only flag can't be used with --target"))
+		}
+		if c.flagMode != moveDefaultMode {
+			return fmt.Errorf(i18n.G("The --mode flag can't be used with --target"))
 		}
```
Comment (stgraber) [inline lxc/move.go]: line break

File: lxc/move.go [hunk 5]
```diff
@@ -140,18 +140,18 @@ func (c *cmdMove) Run(cmd *cobra.Command, args []string) error {
...
+	if c.flagTarget != "" {
+		if c.flagStateless {
+			return fmt.Errorf(i18n.G("The --stateless flag can't be used with --target"))
 		}
-
-		if moved {
-			return nil
+		if c.flagContainerOnly {
+			return fmt.Errorf(i18n.G("The --container-only flag can't be used with --target"))
+		}
```
Comment (stgraber) [inline lxc/move.go]: line break

File: lxc/move.go [hunk 6]
```diff
@@ -140,18 +140,18 @@ func (c *cmdMove) Run(cmd *cobra.Command, args []string) error {
...
-	// that the container is not running.
-	if sourceRemote == destRemote && c.flagTarget != "" {
-		moved, err := maybeMoveCephContainer(conf, sourceResource, destResource, c.flagTarget)
-		if err != nil {
-			return err
+	// cluster member to another, let's use the dedicated API.
+	if c.flagTarget != "" {
+		if c.flagStateless {
+			return fmt.Errorf(i18n.G("The --stateless flag can't be used with --target"))
 		}
```
Comment (stgraber) [inline lxc/move.go]: line break

File: lxd/container_post.go [hunk 7]
```diff
@@ -136,18 +160,35 @@ func containerPost(d *Daemon, r *http.Request) Response {
...
 			}
+
+			// If this is not a ceph-based container, make sure
+			// that the source node is online, and we didn't get
+			// here only to handle the case where the container is
+			// ceph-based.
+			if sourceNodeOffline {
+				err := fmt.Errorf("The cluster node hosting the container is offline")
+				return SmartError(err)
+			}
```
Comment (stgraber) [inline lxd/container_post.go]: line break

File: lxd/container_post.go [hunk 8]
```diff
@@ -203,12 +244,153 @@ func containerPost(d *Daemon, r *http.Request) Response {
...
+
+		// First make a copy on the new node of the container to be moved.
+		entry, _, err := source.GetContainer(oldName)
+		if err != nil {
+			return errors.Wrap(err, "Failed to get container info")
+		}
+
+		args := lxd.ContainerCopyArgs{
+			Name:          destName,
+			Live:          true,
```
Comment (stgraber) [inline lxd/container_post.go]: Hmm, Live should be set to False here (just not set) as we're not doing live migration.

File: lxd/container_post.go [hunk 9]
```diff
@@ -203,12 +244,153 @@ func containerPost(d *Daemon, r *http.Request) Response {
...
+			ContainerOnly: false,
```
Comment (stgraber) [inline lxd/container_post.go]: That's the default so probably not worth setting at all.

File: test/suites/clustering.sh [hunk 10]
```diff
@@ -277,9 +277,13 @@ test_clustering_containers() {
...
 
   # Move the container on node3 to node1, using a client connected to
-  # node2.
+  # node2 and a different container name than the original one.
   LXD_DIR="${LXD_TWO_DIR}" lxc move bar egg --target node1
   LXD_DIR="${LXD_ONE_DIR}" lxc info egg | grep -q "Location: node1"
+
+  # Move back to node3 the container on node1, keeping the same name.
+  LXD_DIR="${LXD_TWO_DIR}" lxc move egg --target node3
+  LXD_DIR="${LXD_ONE_DIR}" lxc info egg | grep -q "Location: node3"
```
Comment (stgraber) [inline test/suites/clustering.sh]: Might be worth checking volatile.apply_template in this test too

File: doc/rest-api.md (hunk 1 above)
Comment (freeekanayaka) [reply doc/rest-api.md]: Fixed.

File: doc/rest-api.md (hunk 2 above)
Comment (freeekanayaka) [reply doc/rest-api.md]: Nothing to amend here afaics, as it's already "member".

File: doc/rest-api.md (hunk 3 above)
Comment (freeekanayaka) [reply doc/rest-api.md]: Fixed.

File: lxc/move.go (hunk 4 above)
Comment (freeekanayaka) [reply lxc/move.go]: Fixed.

File: lxd/container_post.go (hunk 8 above)
Comment (freeekanayaka) [reply lxd/container_post.go]: Fixed.

File: lxc/move.go (hunk 5 above)
Comment (freeekanayaka) [reply lxc/move.go]: Fixed.

File: lxc/move.go (hunk 6 above)
Comment (freeekanayaka) [reply lxc/move.go]: Fixed.

File: lxd/container_post.go (hunk 7 above)
Comment (freeekanayaka) [reply lxd/container_post.go]: Fixed.

File: lxd/container_post.go (hunk 9 above)
Comment (freeekanayaka) [reply lxd/container_post.go]: Fixed. Dropped "Live" too, with the same rational.

File: test/suites/clustering.sh (hunk 10 above)
Comment (freeekanayaka) [reply test/suites/clustering.sh]: Fixed.

File: lxd/container_post.go [hunk 11]
```diff
@@ -136,18 +160,36 @@ func containerPost(d *Daemon, r *http.Request) Response {
...
 			if pool.Driver == "ceph" {
 				return containerPostClusteringMigrateWithCeph(d, c, name, req.Name, targetNode)
 			}
+
+			// If this is not a ceph-based container, make sure
+			// that the source node is online, and we didn't get
+			// here only to handle the case where the container is
+			// ceph-based.
+			if sourceNodeOffline {
+				err := fmt.Errorf("The cluster node hosting the container is offline")
```
Comment (stgraber) [inline lxd/container_post.go]: s/node/member/

File: lxd/container_post.go [hunk 12]
```diff
@@ -203,12 +245,151 @@ func containerPost(d *Daemon, r *http.Request) Response {
...
+	})
+	if err != nil {
+		return SmartError(err)
+	}
+
+	run := func(*operation) error {
+		// Connect to the source host, i.e. ourselves (the node the container is running on).
+		source, err := cluster.Connect(sourceAddress, cert, false)
+		if err != nil {
+			return errors.Wrap(err, "Failed to connect to local node")
```
Comment (stgraber) [inline lxd/container_post.go]: Failed to connect to source cluster member

File: lxd/container_post.go [hunk 13]
```diff
@@ -203,12 +245,151 @@ func containerPost(d *Daemon, r *http.Request) Response {
...
+		}
+
+		// Connect to the destination host, i.e. the node to migrate the container to.
+		dest, err := cluster.Connect(targetAddress, cert, false)
+		if err != nil {
+			return errors.Wrap(err, "Failed to connect to local node")
```
Comment (stgraber) [inline lxd/container_post.go]: Failed to connect to destination server

File: client/lxd_containers.go [hunk 14]
```diff
@@ -280,8 +280,18 @@ func (r *ProtocolLXD) CopyContainer(source ContainerServer, container api.Contai
...
 	}
 
+	sourceInfo, err := source.GetConnectionInfo()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get source connection info: %v", err)
+	}
+
+	destInfo, err := r.GetConnectionInfo()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get source connection info: %v", err)
```
Comment (stgraber) [inline client/lxd_containers.go]: s/source/destination/

File: lxd/container_post.go (hunk 11 above)
Comment (freeekanayaka) [reply lxd/container_post.go]: Fixed.

File: lxd/container_post.go (hunk 12 above)
Comment (freeekanayaka) [reply lxd/container_post.go]: Fixed.

File: client/lxd_containers.go (hunk 14 above)
Comment (freeekanayaka) [reply client/lxd_containers.go]: Fixed.

**Comment by laralar** [comment]
Sorry to bother with this again.. but it is kind of annoying to have to unattach the container from MAAS, and then attach it again

root@node48:~# lxc move blade-dn48 node47:
Error: Failed container creation: {"mac_addresses": ["MAC address 00:16:3e:ab:28:a4 already in use on blade-dn48."], "hostname": ["Node with this Hostname already exists."]}
root@node48:~# lxc move blade-dn48 node47: -p default-nomaas
Transferring container: blade-dn48: 44.79MB (10.58MB/s)

I have solved the issue by adding a "default-nomaas" profile

But I believe that the move command should behave the same if moving locally or to another node.  (same as the cluster implementation)

I have read before that the implementations are different, but behavior should be the same

If I move the MAAS attached container locally, there is no error, but if I move it to another node, there is error.  can't the API implementation unattach the container from MAAS, move it, and the re-attache it?

**Comment by stgraber** [comment]
@laralar what LXD version is that?

**Comment by stgraber** [comment]
Ah, right, you're moving between two nodes which aren't in the same cluster... LXD could detach and re-attach but there's no guarantee that you'll get the same IP address when that happens.

**Comment by laralar** [comment]
Thank you for the quick response

I'm using the latest snap version which I think is 3.5 when I had this.

In don't think the IP address would be an issue since I'm moving a container and the MAC address before and after is the same, since MAAS manages the assignment per MAC address ( even though you could specify an IP address in the device property of the container, if inside the container another static IP address is specified, that, of course it's the expected behavior, takes priority over the container device IP address property )

**Comment by stgraber** [comment]
There is no MAAS API that lets you move a sub-device from one machine to another, so the only way to move a container from one host to another in MAAS is to delete the container and create it again on the target host.

As that means MAAS will be told to forget the container's MAC, so will the DHCP allocation and the container IP.

**Comment by laralar** [comment]
OK.. thanks for the clarification

I haven't had issues with change of IP Address on MAAS allocated IPs.. since I am doing the same, (detaching it from MAAS, and re-attaching it late on),, the IP remains the same.  Even though there was an IP change issue, it shouldn't be a problem since when you are using DHCP you shouldn't rely on the IP Address assigned, but the domain name, since it is part of what happens when you are using DHCP, right?

Could it be because even though MAAS forgets the IP, the dhcpd deamon still has the lease information in the leases DB file, or also the container once it boots  (I have seen old IP Addresses and MAC information in the dhcp leases file from the MAAS server that are not even available (e.g. a container is offline for a looong time and if the IP address is not used yet, the container regains the same address)

I even had to manually delete some known address in MAAS manually using the API command line for some IPs that were used back in July for once, and not used any more, so that I could assign them statically to a host.

**Comment by laralar** [comment]
Of course... if I move the container detaching it.. and wait a looong time before attaching it (what I am doing now), and there are few available IP Addresses in the DHCP lease and the lease has expired, the container will definetely get another IP address if the previous one was assigned to another system

**Comment by laralar** [comment]
I just saw this is a pull request and it is merged,, should I open another issue for what I described as the "same behavior for move between ndoes for MAAS attached containers"?

**Comment by stgraber** [comment]
Within a cluster there may be something we can do, outside of a cluster, I expect this would be closed as not fixable due to how migration works.

The way the client runs `lxc move` is by first copying your source container to the target and then deleting it at the source, so it's impossible to avoid the situation where two containers exist, connected to the same MAAS with the same MAC.

That's actually why we added the `-d` option to `lxc move` so that you can unset properties like maas.ipv4.subnet during the copy.

In the cluster case, we have an internal move API which could be made to handle this case, though again with no guarantee that MAAS won't just re-use your lease halfway through the copy.

**Comment by laralar** [comment]
fair enough.. thanks.   

the -d property works great when the container does have the MAAS settings or IP address in the device.   nt when those setting are for example in the default profile that is the default

thanks

**Comment by laralar** [comment]
do you mean that within the cluster may work, even if the clster is connected to MAAS, but not if the container is connected to MAAS and not in a cluster, right?

**Comment by stgraber** [comment]
Correct, the cluster has a different move API which would let it do the device removal and re-add during the copy so as to never have the same mac and ip twice. That's because inside a cluster we can't do the usual copy+delete workflow as that'd result in the same container existing twice in the database (which isn't allowed).

**Pull request merged**